Cassandra-15971 Improve FQL doc page

polandll opened the pull request:
I've rewritten this page to have better organization and flow. Needs review.

polandll committed:
rewrite FQL

polandll committed:
improve FQL page

polandll committed:
make edits per Ekaterina

Deepak-Vohra commented:
Please verify that all sections that need to be in code font are in code font. As example, the following code snippet is not completely in code font.  
```
replay \
--keyspace querylogkeyspace --results /cassandra/fql/logs/results/replay --store-queries /cassandra/fql/logs/queries/replay -- target 3.91.56.164 /tmp/cassandrafullquerylog 
```

polandll committed:
fix codeblock per Deepak-Vohra comment

polandll commented:
Thanks for catching that codeblock. I forgot to indent. :-) I spend much more time in Asciidoc than reStructured text. I'm made the fix.

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -14,174 +14,169 @@
+- Introduces new ``fqltool`` that currently can ``Dump`` the binary logs to a readable format. Other options are ``Replay`` and ``Compare``.
 
-Cassandra 4.0 has a binary full query log based on Chronicle Queue that can be controlled using ``nodetool enablefullquerylog``, ``disablefullquerylog``, and ``resetfullquerylog``. The log contains all queries invoked, approximate time they were invoked, any parameters necessary to bind wildcard values, and all query options. A readable version of the log can be dumped or tailed using the new ``b ...
+FQL logs all Cassandra Query Language (CQL) requests, both events that modify the data and those that query. 
```
> FQL logs all Cassandra Query Language (CQL)

should be

> FQL logs all successful Cassandra Query Language (CQL)

Checking the code, bad queries (ones which don't parse or cause an exception) will not be logged.

dcapwell replied:
see you say this later with ` Failed or timed-out queries are not logged.`, still good to be clear

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -14,174 +14,169 @@
+^^^^^^^^^^^^^^^
 
-::
+The ``archive_command`` option sets the user-defined archive script to execute on rolled log files. 
```
should state that when not defined, files are deleted

dcapwell replied:
Default is `""` which then maps to `org.apache.cassandra.utils.binlog.DeletingArchiver`

```
BinLogArchiver archiver = Strings.isNullOrEmpty(archiveCommand) ? new DeletingArchiver(maxLogSize) : new ExternalArchiver(archiveCommand, path, maxArchiveRetries);
```

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -14,174 +14,169 @@
+For example: ``max_archive_retries: 10``
 
-The ``full_query_log_dir`` setting is not within the ``full_query_logging_options`` but still is for full query logging.
+FQL can also be configured using ``nodetool` when enabling the feature, and will override any values set in the `cassandra.yaml` file, as discussed in the next section.
```
nodetool has 2 ` upfront, should be 1

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -239,74 +234,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
+The ``fqltool`` command is used to view (dump), replay, or compare logs.
+``fqltool dump`` converts the binary log files into human-readable format; only the log directory must be supplied as a command-line option.
+
+``fqltool replay`` (CASSANDRA-14618) enables replay of logs. 
```
should this link to `https://issues.apache.org/jira/browse/CASSANDRA-14618`?  Previous examples did

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -239,74 +234,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
+
+``fqltool replay`` (CASSANDRA-14618) enables replay of logs. 
+The command can run from a different machine or cluster for testing, debugging, or performance benchmarking. 
+The command, run on the same node on which the logs are generated can recreate a dropped database object.
```
> The command, run on the same node on which the logs are generated can recreate a dropped database object

I don't know what this means, what are you trying to say?

polandll replied:
I didn't write that text originally. I've revised to: The command can also be used to recreate a dropped database object.

dcapwell reviewed (commented):
Overall LGTM, only small comments

polandll committed:
revise per dcapwell comments

polandll commented:
Thanks @dcapwell - I've incorporated your comments!

smiklosovic commented on doc/source/new/fqllogging.rst:
```diff
@@ -239,74 +235,186 @@ The syntax of the nodetool enablefullquerylog command is as follows:
+
+::
+
+  fqltool replay [--keyspace <keyspace>] [--results <results>]
```
@polandll you will likely have merge conflicts somewhere along these lines as this was just merged recently.

https://github.com/apache/cassandra/commit/00c22232d96b6e592fb1d1f32a6b649b49eceb34

polandll replied:
Thanks for the heads-up. I resolved the conflict.

polandll committed:
Merge branch 'trunk' into CASSANDRA-15971-fql-docs

polandll commented:
@dcapwell - I've added the --replay-ddl-statements flag to the replay command. I added a note to indicate that replay is generally done between clusters, but left the example using a single cluster, just to simplify. I hope that this change meets with your approval.

polandll committed:
fix example, add replay-ddl-statement

polandll committed:
Merge branch 'CASSANDRA-15971-fql-docs' of https://github.com/polandll/cassandra into CASSANDRA-15971-fql-docs

smiklosovic commented:
@dcapwell it would be really nice if this is in https://github.com/apache/cassandra/pull/721

I am introducing there getfullqerylog command in nodetool and it touches the documentation yet again so it would have to be added there etc.

dcapwell commented:
working on it now

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -498,198 +558,31 @@ The full query logs get listed. Truncated output is as follows:
+ $ fqltool replay \
+ --keyspace querylogkeyspace --results /cassandra/fql/logs/results/replay \
+ --store-queries /cassandra/fql/logs/queries/replay \
+ -- target 3.91.56.164 \
```
extra space

dcapwell commented on doc/source/new/fqllogging.rst:
```diff
@@ -498,198 +558,31 @@ The full query logs get listed. Truncated output is as follows:
-As an example of using ``fqltool replay``, drop a keyspace.
-
-::
+ $ fqltool replay \
```
since we do drop keyspace, will need `--replay-ddl-statements`.

dcapwell reviewed (commented):
Overall LGTM.  2 small comments but I can fix that on merge.
